pycbc/inference/gelman_rubin.py

```diff
@@ -0,0 +1,174 @@
+
+    # get number of chains, parameters, and iterations
+    chains = numpy.array(chains)
+    nchains, nparameters, niterations = chains.shape
+
+    # calculate the covariance matrix for each chain
+    # this will have shape (nchains, nparameters, nparameters)
+    chains_covs = numpy.array([numpy.cov(chain) for chain in chains])
+    if nparameters == 1:
+        chains_covs = chains_covs.reshape((nchains, 1, 1))
+
+    # calculate W
+    # this will have shape (nparameters, nparameters)
+    w = numpy.zeros(chains_covs[0].shape)
+    for i, row in enumerate(chains_covs[0]):
+        for j, _ in enumerate(row):
+            w[i, j] = numpy.mean(chains_covs[:, i, j])
+    if nparameters == 1:
+        w = w.reshape((1, 1))
+
+    # calculate B
+    # this will have shape (nparameters, nparameters)
+    means = numpy.zeros((nparameters, nchains))
+    for i, chain in enumerate(chains):
+        means[:, i] = numpy.mean(chain, axis=1).transpose()
+    b = niterations * numpy.cov(means)
+    if nparameters == 1:
+        b = b.reshape((1, 1))
+
+    # get diagonal elements of W and B
+    # these will have shape (nparameters)
+    w_diag = numpy.diag(w)
+    b_diag = numpy.diag(b)
+
+    # get variance for each chain
+    # this will have shape (nparameters, nchains)
+    var = numpy.zeros((nparameters, nchains))
+    for i, chain_cov in enumerate(chains_covs):
+        var[:, i] = numpy.diag(chain_cov)
+
+    # get mean of means
+    # this will have shape (nparameters)
+    mu_hat = numpy.mean(means, axis=1)
+
+    # get variance of variance
+    # this will have shape (nparameters)
+    s = numpy.var(var, axis=1)
+
+    # get factors in variance of V calculation
+    # this will have shape (nparameters)
+    k = 2 * b_diag**2 / (nchains - 1)
+    mid_term = numpy.cov(var, means**2)[nparameters:2 * nparameters, 0:nparameters].T
+    end_term = numpy.cov(var, means)[nparameters:2 * nparameters, 0:nparameters].T
+    wb = niterations / nchains * numpy.diag(mid_term - 2 * mu_hat * end_term)
+
+    # get V
+    # this will have shape (nparameters)
+    v = (niterations - 1.) * w_diag / niterations + (1. + 1. / nchains) * b_diag / niterations
+
+    # get variance of V
+    # this will have shape (nparameters)
+    var_v = ((niterations - 1.)**2 * s + (1. + 1. / nchains)**2 * k + \
+             2. * (niterations - 1.) * (1. + 1. / nchains) * wb) / niterations**2
+
+    # get degrees of freedom
+    # this will have shape (nparameters)
+    dof = (2. * v**2) / var_v
+
+    # more degrees of freedom factors
+    # this will have shape (nparameters)
+    df_adj = (dof + 3.) / (dof + 1.)
+
+    # estimate R
+    # this will have shape (nparameters)
+    r2_fixed = (niterations - 1.) / niterations
+    r2_random = (1. + 1. / nchains) * (1. / niterations) * (b_diag / w_diag)
+    r2_estimate = numpy.sqrt(r2_fixed + r2_random)
+
+    # PSRF
+    # this will have shape (nparameters)
+    psrf = r2_estimate * df_adj
```

Comment:
I was looking up the math of Gelman Rubin statistics in resources available on the internet, and it looks like there should be a sqrt over `df_adj` here. So, I think this would be `psrf = r2_estimate * sqrt(df_adj)`